### ISSUE TITLE
[Request] Support Panopticon (Discord IRC Styled Logger)

### ISSUE THREAD

**Issue by dubyadud** [tommyrot/superseriousstats#11]
[Panopticon](https://github.com/ihaveamac/panopticon) is a [Discord](https://discordapp.com) logger.  If the two could work together, it would open the doors to Discord + SuperSeriousStats!

Cheers, yay or neigh.

**Comment by tommyrot** [comment]
Can Panopticon log activity to a per-day file having the year, month and day in the filename? That's something sss requires.

We also need their log syntax, either extracted from their source/dependency, or by means of going over some of their logs which contain all different line types and events (actions, kicks, topics, joins, etc). The less i have to do to find this out, the more likely it is i'll write a parser for it.

**Comment by ihaveamac** [comment]
(note: i'm the current maintainer of panopticon above, and i'm willing to make any changes that are necessary)
> Can Panopticon log activity to a per-day file having the year, month and day in the filename? That's something sss requires.

It does this right now, in the form of `Sever Name-serverid/#channel-name-channelid/yyyy-mm-dd.log`  
(example: `rOverwatch-94882524378968064/#general-discussion-94882524378968064/2017-10-13.log`)

(This does mean if the server or channel names change, the logs will become separated. I could change it to just save based on ID, but I'm not sure at the moment)

> We also need their log syntax, either extracted from their source/dependency, or by means of going over some of their logs which contain all different line types and events (actions, kicks, topics, joins, etc). The less i have to do to find this out, the more likely it is i'll write a parser for it.

Usernames go like `user name#dscm` (dscm = 4-number discriminator, to separate the user from others with the same username).

The current message syntax goes like this (the b64 id can begin with E: meaning an edit):
```
[b64-encoded id] [hh:mm:ss] <user name#dscm> Message
(newline) Newline, if it has one
(attach) URL to attachment
```
Example:
```
[AQCCe0JCBwY=] [16:25:46] <ihaveahax#1684> Message 
[AAACjkpCBwY=] [16:25:55] <ihaveahax#1684> Message
(newline) with a newline 
[DADAMVBCBwY=] [16:26:00] <ihaveahax#1684> Message (without edit) 
[E:DADAMVBCBwY=] [16:26:03] <ihaveahax#1684> Message (with edit) 
[FQAAR3NCBwY=] [16:26:36] <ihaveahax#1684> Message
(newline) with newline and attachment 
(attach) https://cdn.discordapp.com/attachments/327459864945229824/434388951944396820/StartupMovie.gif
```
The ID can be ignored, that's only used for things like the Discord API.

At the moment, panopticon doesn't log events like joins, leaves, topic changes, etc. Kicks and bans may not happen since seeing those requires extra permissions that the bot may not have, but joins + leaves can to a separate log file (since this is a per-server thing and not per-channel), and topic changes could be saved to the channel's log file.

**Comment by tommyrot** [comment]
I see a couple of things that don't go well with how sss parses logfiles, perhaps these can be worked out. First, newlines in a message will break things as general demarcation in logs is the newline. So for example:

```
[12:01] <user> hello
this is a test
```

The above will succesfully parse the "hello" line but drops the "this is a test" one because it doesn't match anything (no msg id _although ignored in the example_, no time, no username). Unless.. the user crafts their lines after newlines in such a way that it **does** match an event or line type and inject unwanted kicks (and such) into your stats.
So these lines are a problem, both to ignore and to parse. Perhaps they can be prefixed with the exact same msg id + time + username as the first line by that user. The sss parser is pretty dumb, it can't prefix that information itself because it can't tell if the line is actually part of valid input by the user or indeed something that should be ignored.

The second thing i noticed are the edits (E:). The msg id appears to remain the same but it will require more logic to handle by the parser. Questions arise; the original message could be written hours ago, processing this edit will impact the timeline/flow of the chat as represented by the stats. Dropping the original after an edit is observed won't be possible if the original has already been processed by an earlier sss run (depending on the frequency, could be every hour even less). I think the best approach here is to ignore edits and go with the original message.

**Comment by ihaveamac** [comment]
About newlines: panopticon currently inserts `(newline)` at the beginning of every line after the first to try and make it easier to parse, and to avoid new lines from faking entries. If it's still too hard to parse, I can make newlines use the same id+time+name as the first line.
```
[AABEap13DwY=] [01:26:44] <ihaveahax#1684> some message 
[AADEeMB3DwY=] [01:27:20] <ihaveahax#1684> another message
(newline) [AABEap13DwY=] [01:26:44] <user#1234> part of the same message above
```

About edits: I think ignoring them would probably be the best idea. Edits that change the entire message don't tend to happen often. Plus, edits for really old messages wouldn't be caught by the logger, since it stores a cache of the most recent messages.

**Comment by tommyrot** [comment]
Ah i didn't realise (newline) was literal. Well that should be workable i think, i'll have a look coming days. Best to not change your logging syntax if not needed because surely there are other tools relying on it as well.
